Add first blog post for temporal_rs and general Boa Temporal work (#195)
boa-dev/boa-dev.github.io: blog/temporal-introduction.md

Reviewer: Question: shouldn't this be prefix by datetime like other posts are? `2025-06-15-temporal-introduction.md`

Author: Yeah, I can add that. I think it may also need a datetime in the header too

Author: Oh, the datetime in the file name dictates the header value. Hmmm, I guess that helps keep things organized.

---
authors: boa-dev
---

# Implementing Temporal, the new date/time API for JavaScript (and Rust!)

This will be a series of posts primarily about implementing a new

> JavaScript feature in Rust, specifically the new date/time builtin:

Reviewer: ```suggestion
JavaScript feature in Rust, specifically the new date/time built-in:
```

Temporal. We'll be going over the general implementation of Temporal in
Boa as well as the crates supporting that implementation.

Why should you care? Well we are not only implementing it for
JavaScript, but Rust as well ... more on that in a bit.

First, an aside!

## What even is Temporal?

Temporal is a modern API for handling date/time in a calendar and time
zone aware manner that includes nine objects with over 200+ methods.

> In JavaScript, Temporal is a global builtin namespace object that
> includes each of these nine builtins:

Reviewer: ```suggestion
In JavaScript, Temporal is a global built-in namespace object that
includes each of these nine built-ins:
```

> - Temporal.Now
> - Temporal.PlainDate
> - Temporal.PlainTime
> - Temporal.PlainDateTime
> - Temporal.ZonedDateTime
> - Temporal.Duration
> - Temporal.Instant
> - Temporal.PlainYearMonth
> - Temporal.PlainMonthDay

Reviewer: Remark:

```suggestion
- `Temporal.Now`
- `Temporal.PlainDate`
- `Temporal.PlainTime`
- `Temporal.PlainDateTime`
- `Temporal.ZonedDateTime`
- `Temporal.Duration`
- `Temporal.Instant`
- `Temporal.PlainYearMonth`
- `Temporal.PlainMonthDay`
```

But to be honest, this post isn't meant to give an overview of Temporal
and its general API. If Temporal is news to you and you are interested
in learning more, feel free to check out the phenomenal [MDN
documentation][mdn-temporal].

## Back on track

Being a JavaScript engine / interpreter, implementing the ECMAScript
specification is our raison d'être. One of our primary goals is to
implement the specification to high degree of conformance. This is even
more so important because Temporal represents roughly 7-8% of the
current conformance test suite (~4000 of the ~50,000 tests).

When the PR of the first prototype of Temporal for Boa was submitted a
few things became evident.

1. Date/Time is a complicated beast (duh)
2. There's room for optimization and improvement
3. This would be handy to have in Rust

So after the prototype was merged, the prototype was pulled out of Boa's

> internal builtins and externalized into its own crate,

Reviewer: ```suggestion
internal built-ins and externalized into its own crate,
```

[`temporal_rs`][temporal-rs-repo], which then first landed behind an

> experimental flag in Boa v0.18.

Reviewer 2: ```suggestion
So, after the prototype was merged, the prototype was pulled out of Boa's
internal builtins and externalized into its own crate,
[temporal_rs][temporal-rs-repo], which landed behind an
experimental flag in Boa v0.18 at first.
```

primitive to handle casting and constraining, respectively, to adapt
values for use with a typed API.

For instance, in Boa, we heavily lean into using the below patterns:

```rust
// (Note: this is abridged for readability)

// FiniteF64 usage
let number: f64 = js_value.to_number(context)?;
let finite_f64: FiniteF64 = FiniteF64::try_from(number)?;
let year: i32 = finite_f64.as_integer_with_truncation::<i32>();

// FromStr usage with `get_option`
let options_obj: &JsObject = get_options_object(&js_value)?;
let overflow: Option<ArithmeticOverflow> = get_option::<ArithmeticOverflow>(
    &options_obj,
    js_string!("overflow"),
    context
)?;
```

This is the core glue between Boa and the `temporal_rs` API that we will
be going over below.

## Implementing constructors

There are a variety of ways to construct a core component like
`PlainDate`, and that stems from the core constructor for each of the
core components `new_with_overflow`:

```rust
impl PlainDate {
    pub fn new_with_overflow(year: i32, month: u8, day: u8, calendar: Calendar, overflow: ArithmeticOverflow) -> Result<Self> {
        // Create PlainDate
    }
}
```

> This function supports the baseline construction of Temporal builtins,

Reviewer: ```suggestion
This function supports the baseline construction of Temporal built-ins,
```

which takes the usual year, month, day, alongside a calendar and also an
overflow option to constrain or reject based on whether the provided
values are in an expected range.

However, we can better express this in Rust with common `try_` prefix
notation.

```rust
impl PlainDate {
    pub fn new(year: i32, month: u8, day: u8, calendar: Calendar) -> Result<Self> {
        Self::new_with_overflow(year, month, day, calendar, ArithmeticOverflow::Constrain)
    }

    pub fn try_new(year: i32, month: u8, day: u8, calendar: Calendar) -> Result<Self> {
        Self::new_with_overflow(year, month, day, calendar, ArithmeticOverflow::Reject)
    }
}
```

These methods are flexible enough and allow full coverage of the
Temporal specification.

For instance, take the below constructor:

```js
const plainDate = Temporal.PlainDate(2025, 6, 9);
```

Can easily be conveyed in Rust as:

```rust
let plain_date = PlainDate::try_new(2025, 6, 9, Calendar::default())?;
```

We actually learn some interesting things immediately about the
JavaScript API from looking at the `temporal_rs` API:

> 1. Temporal.PlainDate constructor can throw.

Reviewer: ```suggestion
1. `Temporal.PlainDate` constructor can throw.
```

2. When the calendar is omitted, the default calendar is used (this will
   default to the `iso8601` calendar)

Of course, if you somewhat prefer the brevity of the JavaScript API and

> don't want to list the default Calendar, `temporal_rs` provides the

Reviewer: ```suggestion
don't want to list the default `Calendar`, `temporal_rs` provides the
```

additional constructors `new_iso` and `try_new_iso`.

> ## Let's discuss Now

Reviewer: Maybe?
```suggestion
## Let's discuss `Now`
```

> Colonel Sandurz: Now. You're looking at now, sir. Everything that
> happens now, is happening now.<br></br> Dark Helmet: What happened to
> then?<br></br> Colonel Sandurz: We passed then.<br></br> Dark Helmet:
> When?<br></br> Colonel Sandurz: Just now. We're at now now.<br></br>
> Dark Helmet: Go back to then.<br></br> Colonel Sandurz: When?<br></br>
> Dark Helmet: Now.<br></br> Colonel Sandurz: Now?<br></br> Dark Helmet:
> Now.<br></br> Colonel Sandurz: I can't.<br></br> Dark Helmet:
> Why?<br></br> Colonel Sandurz: We missed it.<br></br> Dark Helmet:
> When?<br></br> Colonel Sandurz: Just now.<br></br> Dark Helmet: When
> will then be now?<br></br> -- <cite>Spaceballs, 1987</cite>

`Temporal.Now` is an incredibly strange type, yet nevertheless
important. It is the object from which the current instant can be
measured and mapped into any of the Temporal components.

> In JavaScript, this type has no [[Construct]] or [[Call]] internal

Reviewer: ```suggestion
In JavaScript, this type has no `[[Construct]]` or `[[Call]]` internal
```

method, which is a fancy way to say that Now has no constructor and
cannot be called directly.

> Instead, Now is used primarily as a namespace for it's methods.

Reviewer: ```suggestion
Instead, Now is used primarily as a namespace for its methods.
```

It turns out that while property bags in JavaScript can have various
fields set, there is still a general shape for the fields that can be
provided and validated in Temporal.

So in order to support this in `temporal_rs`, a "partial" component
exists for each of the components that can then be provided to that
components `from_partial` method.

With this, we have fully implemented support for the `from` method in
`temporal_rs`:

```rust
use core::str::FromStr;
use temporal_rs::{PlainDate, PlainDateTime, partial::PartialDate};
let pdt = PlainDateTime::try_new_iso(2025, 1, 1)?;
// We can use the `PlainDateTime` (`ZonedDateTime` / `PlainDate` are also options).
let pd_from_pdt = PlainDate::from(pdt);
// We can use a `str`.
let pd_from_string = PlainDate::from_str("2025-01-01")?;
// We can use a `PartialDate`.
let pd_from_partial = PlainDate::from_partial(PartialDate {
    year: Some(2025),
    month: Some(1),
    day: Some(1),
    ..Default::default()
});
```

**NOTE:** there may be updates to `PartialDate` in the future (see
[boa-dev/temporal #349](https://github.com/boa-dev/temporal/issues/349)
for more information).

## Elephant in the room: time zones

> So far we have not dicussed time zones, and -- suprise! -- we aren't

Reviewer: ```suggestion
So far we have not discussed time zones, and -- surprise! -- we aren't
```

Author: Oops lol I've looked at this line so many times too.

we'll hopefully go over everyone's favorite subject, time zones; and
answer the question that some of you may have if you happen to take a
glance at `temporal_rs`'s docs or try out our `no_std` support: what in
the world is a provider API?

## Conclusion

In conclusion, we're implementing Temporal in Rust to support engine
implementors as well as to have the API available in native Rust in
general.

Boa currently sits at a [90% conformance rate][boa-test262] for Temporal
completely backed by `temporal_rs` v0.0.8, and we're aiming to be 100%
conformant before the end of the year.

If you're interested in trying Temporal using Boa, you can use it in
Boa's CLI or enable it in `boa_engine` with the `experimental` flag.

Outside of Boa's implementation, `temporal_rs` has implemented or
supports the implementation for a large portion of the Temporal's API in
native Rust.

If you're interested in trying out `temporal_rs`, feel free to add to
your project with Cargo:

```bash
cargo add temporal_rs
```

or by adding the below to your `Cargo.toml`

```toml
temporal_rs = "0.0.9"
```

A FFI version of temporal is also available for C and C++ via

> temporal_capi.

Reviewer: ```suggestion
`temporal_capi`.
```

## General note on API stability

While the majority of the APIs discussed above are expected to be mostly
stable. Temporal is still a stage 3 proposal that is not fully accepted
into the ECMAScript specification. Any normative change that may be made
upstream in the ECMAScript or ECMA402 specification will also be
reflected in `temporal_rs`.

There are also a few outstanding issues with changes that may reflect in
the API.

> 1.  Duration's inner repr and related ctors

Reviewer: ```suggestion
1.  Duration's inner repr and related constructors.
```
